### DISCUSSION TITLE
What defines an "album" when importing tracks?

### DISCUSSION THREAD

**Discussion by sjlongland** [beetbox/beets#4468] [Q&A] [answered]
Hi all,

Silly question… what defines an "album" in `beets`?  I'm new to `beets`, having managed my music collection (over 7000 tracks) using self-built Perl scripts for many years, then later moving the code across to Python.

My day-to-day repository is Ogg/Vorbis, but I'm building up an archive which uses FLAC.  The source material is a combination of purchased digital download and rips from CDs, DVDs and records (mostly 33⅓RPM LPs, but there is also one 45RPM).  The albums include both releases from artists as well as compilations (multi-artist).

All tracks are tagged with, at minimum (there are other tags in there too like replaygain, scoring and some non-standard tags):
* Title
* Variant (describes things like "live" recordings, remastered versions, singles, etc, rather than polluting `TITLE` with this information)
* Artist
* Album Artist (`ALBUMARTIST` Vorbis tag)
* Album
* Disc number (where applicable)
* Side (for LPs; where applicable -- not used for CDs, DVDs and digital download)
* Track number (relative to the start of the "album")
* Track offset (relative to the start of a disc or side of disc)
* UUID

Most are now tagged with cover art as well.  I figured I'd use `beets` to manage this archive repository.  There's still a lot I need to add to the archive, mostly LPs that need to go back on the turntable as I didn't keep the uncompressed recording when I last ripped them, but right now I have 6374 FLAC files, which all have correct metadata.  (I have them as Ogg/FLAC right now as it was easier to work with using `tagpy`, but I'm slowly re-working my tooling to support FLAC directly and converting the files directly to plain FLAC.)

When I import these into `beets` (with the `-A` flag, since they're already tagged), `beets` tells me I have 6374 "albums".  If I tell it to rename everything, it creates new directories, one for each track -- it thinks each and every track is a distinct album!

So an album like [this one](https://musicbrainz.org/release/80defad2-7de4-4815-9f60-8be1bf1a04b7) shows up as 53 separate "albums".

This is the sort of metadata that `beets` would have seen:

```
stuartl@vk4msl-ws:~$ metaflac --list --block-type=VORBIS_COMMENT /mnt/music-archive/by-album-artist/10cc/I\'m\ Not\ In\ Love\ --\ The\ Essential\ 10cc/d1o001t001\ 10cc\ -\ I\'m\ Not\ In\ Love.oga 
METADATA block #1
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 552
  vendor string: reference libFLAC 1.3.3 20190804
  comments: 19
    comment[0]: ALBUM=I'm Not In Love -- The Essential 10cc
    comment[1]: ALBUMARTIST=10cc
    comment[2]: ALBUMS=d1/o014/t014;10cc;10cc Collected,d3/o011/t049;Various;Now That's What I Call 70s
    comment[3]: ALBUMS_INFO_SEP=;
    comment[4]: ALBUMS_SEP=,
    comment[5]: ARTIST=10cc
    comment[6]: ARTISTS=10cc
    comment[7]: ARTIST_SEP=,
    comment[8]: DATE=2016
    comment[9]: DISCNUMBER=1
    comment[10]: FMPS_PLAYCOUNT=15
    comment[11]: FMPS_RATING_AMAROK_SCORE=0.90
    comment[12]: MEDIA=CD
    comment[13]: REPLAYGAIN_TRACK_GAIN=-3.38 dB
    comment[14]: REPLAYGAIN_TRACK_PEAK=0.99103451
    comment[15]: TITLE=I'm Not In Love
    comment[16]: TRACKNUMBER=1
    comment[17]: TRACK_OFFSET=1
    comment[18]: UUID=4b99c249-646e-4c83-be78-ad4f929ee28b
stuartl@vk4msl-ws:~$ metaflac --list --block-type=VORBIS_COMMENT /mnt/music-archive/by-album-artist/10cc/I\'m\ Not\ In\ Love\ --\ The\ Essential\ 10cc/d1o002t002\ 10cc\ -\ Life\ Is\ A\ Minestrone.oga 
METADATA block #1
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 508
  vendor string: reference libFLAC 1.3.3 20190804
  comments: 19
    comment[0]: ALBUM=I'm Not In Love -- The Essential 10cc
    comment[1]: ALBUMARTIST=10cc
    comment[2]: ALBUMS=d2/o004/t020;10cc;10cc Collected
    comment[3]: ALBUMS_INFO_SEP=;
    comment[4]: ALBUMS_SEP=,
    comment[5]: ARTIST=10cc
    comment[6]: ARTISTS=10cc
    comment[7]: ARTIST_SEP=,
    comment[8]: DATE=2016
    comment[9]: DISCNUMBER=1
    comment[10]: FMPS_PLAYCOUNT=9
    comment[11]: FMPS_RATING_AMAROK_SCORE=0.90
    comment[12]: MEDIA=CD
    comment[13]: REPLAYGAIN_TRACK_GAIN=-3.43 dB
    comment[14]: REPLAYGAIN_TRACK_PEAK=1.01868546
    comment[15]: TITLE=Life Is A Minestrone
    comment[16]: TRACKNUMBER=2
    comment[17]: TRACK_OFFSET=2
    comment[18]: UUID=cf049641-cb9a-43e2-a4fd-3bed958a4bfa
```

What is it about these files that makes `beets` think they belong as "separate" albums?

**Comment by arsaboo** [comment, marked as answer]
By default, beets considers every folder as an album. It considers leaf directories as albums (e.g., typical Artist/Album organization).

**Reply by sjlongland** [reply]
Right, and because my files natively are stored by UUID (so those files would have been in `4b99/c249/646e/4c83/be78/ad4f/929e/e28b.oga` and `cf04/9641/cb9a/43e2/a4fd/3bed/958a/4bfa.oga` respectively, the paths given to `metaflac` are in fact symlinks.)

Okay, so is there a way to instead consider the `(ALBUMARTIST, ALBUM)` tuple as an album instead of the directory?  After all, the files are tagged with metadata specifically so that you don't have to parse the directory structure.

**Reply by sjlongland** [reply]
https://beets.readthedocs.io/en/stable/reference/config.html#group-albums seems to suggest a flag that will consider metadata over directory structure.

That said, `group_albums` doesn't in itself say anything about filesystem structure or metadata.  Is this what was being referred to?

**Reply by arsaboo** [reply]
No, each folder (or a subfolder) is considered an album. `group-albums` can use the meta-data to create albums based from a flat structure.

**Reply by sjlongland** [reply]
So how do I disregard the file structure entirely?  "By default" means there's a setting that can be turned off… what is it?

**Reply by arsaboo** [reply]
You can use `--flat` option. This takes all the music files under the directory (recursively) and treats them as a single large album instead of as one album per directory.

**Reply by sjlongland** [reply]
By the sounds of things… I can:

* import each file individually as its own album (because it happens to be in a directory by itself)
* import all files into a single album (using `--flat`)

If there's nothing in between, it sounds like `beets` is not going to work for my use case.

**Reply by sampsyo** [reply]
You can use `--flat` and `--group-albums` together, if that's what you need.
